### ISSUE TITLE
Bug: Plugins that re-export types from utils are still incompatible with `defineConfig` types

### ISSUE THREAD

**Issue by wadefletch** [typescript-eslint/typescript-eslint#11543]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```TypeScript
// config-only typescript issue
```

### ESLint Config

```javascript
import { defineConfig } from "eslint/config";
import pluginImportX from "eslint-plugin-import-x";
import pluginImportZod from "eslint-plugin-import-zod";

export default defineConfig([
  pluginImportX.flatConfigs.recommended, // ts error here
  pluginImportZod.configs.recommended, // basically same ts error here
  {
    parser: "@typescript-eslint/parser",
  }
]);
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Happy typescript.

### Actual Result

```
Type 'PluginFlatConfig' is not assignable to type 'InfiniteArray<ConfigWithExtends>'.
  Type 'PluginFlatConfig' is not assignable to type 'ConfigWithExtends'.
    Types of property 'languageOptions' are incompatible.
      Type 'import("/Users/wadefletcher/Development/Tractorbeam/style-guide/node_modules/.pnpm/@typescript-eslint+utils@8.41.0_eslint@9.34.0_jiti@2.5.1__typescript@5.9.2/node_modules/@typescript-eslint/utils/dist/ts-eslint/Config").FlatConfig.LanguageOptions | undefined' is not assignable to type 'import("/Users/wadefletcher/Development/Tractorbeam/style-guide/node_modules/.pnpm/eslint@9.34.0_jiti@2.5.1/node_modules/eslint/lib/types/index").Linter.LanguageOptions | undefined'.
        Type 'import("/Users/wadefletcher/Development/Tractorbeam/style-guide/node_modules/.pnpm/@typescript-eslint+utils@8.41.0_eslint@9.34.0_jiti@2.5.1__typescript@5.9.2/node_modules/@typescript-eslint/utils/dist/ts-eslint/Config").FlatConfig.LanguageOptions' is not assignable to type 'import("/Users/wadefletcher/Development/Tractorbeam/style-guide/node_modules/.pnpm/eslint@9.34.0_jiti@2.5.1/node_modules/eslint/lib/types/index").Linter.LanguageOptions'.
          Types of property 'parser' are incompatible.
            Type 'LooseParserModule | undefined' is not assignable to type 'Parser | undefined'.
              Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Parser | undefined'.
                Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'ObjectMetaProperties & { parseForESLint(text: string, options?: any): ESLintParseResult; }'.
                  Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; scopeManager?: unknown; services?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): ESLintParseResult; }'.
                    The types of 'parseForESLint(...).ast' are incompatible between these types.
                      Type 'unknown' is not assignable to type 'Program'.ts(2322)
```

### Additional Info

**Potentially Relevant Discussion**
https://github.com/typescript-eslint/typescript-eslint/pull/11475
https://github.com/un-ts/eslint-plugin-import-x/issues/420

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/utils`             | `8.41.0` |
| `TypeScript`                           | `5.9.1` |
| `ESLint`                               | `9.34.0` |
| `node`                                 | `22.11.0` |

**Comment by kirkwaiblinger** [comment]
@wadefletch 

Yeah, so this was a known consequence of the way we went about solving #10899 (https://github.com/typescript-eslint/typescript-eslint/pull/11475), since I didn't end up changing those types at all, just the types of typescript-eslint's exports.

The reason I went with the approach in #11475, which gives extremely imprecise types, is because after months of trying, I was never able to find a nontrivial type that could satisfy both the typescript-eslint and eslint type declarations, part of the problem for us being that we _also_ need to be support eslint v8 with typescript-eslint. 

Depending on their support requirements downstream users may be able to use eslint native types now, and avoid running into the problems we have. Or, plugins may want to follow our approach and use extremely imprecise types rather than exporting `FlatConfig.Plugin`, etc. Not sure I have a one-size-fits-all solution to this. 

Maybe also other team members will have thoughts... @typescript-eslint/triage-team

**Comment by JoshuaKGoldberg** [comment]
> follow our approach and use extremely imprecise types

My intuition is, yes, this is the way. Users generally shouldn't need to reach into specifics of plugins very much. As long as the right configs are exported then >99% of userland configs should be satisfied.

@kirkwaiblinger is there any information being lost in this approach that you think plugins might miss?

**Comment by kirkwaiblinger** [comment]
No, I think that plugins/parsers should be purely black boxes to users, with basically the one exception that other plugins are likely to want to expose the presence of `plugin.configs` to users, to allow them to set configs (which we don't need to do, since we have the separate `tseslint.configs` entry point).

The main risk would just be making the types so vague that they can be assigned in places they shouldn't go in `defineConfig()`, but idk if that's a real problem.

**Comment by Zamiell** [comment]
I am adding here for searchability purposes that any custom ESLint plugin that uses the handy `ESLintUtils.RuleCreator` helper function is currently in a bugged state with regards to `defineConfig`, due to the underlying problem described in the above OP.

For now, end-users of such plugins have two options:

1) Use `defineConfig` and then slap a `@ts-expect-error https://github.com/typescript-eslint/typescript-eslint/issues/11543` to suppress the type error.
2) Continue to use the deprecated `tseslint.config` helper function and suppress the `@typescript-eslint/no-deprecated` ESLint rule.

**Comment by kirkwaiblinger** [comment]
Is there anything that's actually actionable for us here? Maybe add a blurb to https://typescript-eslint.io/developers/eslint-plugins/ ?

**Comment by ehoogeveen-medweb** [comment]
If there's a particular change that plugin developers need to make to fix this problem for end users, I think adding that solution somewhere would be very helpful. I think some developers are taking a wait and see approach right now, not sure whether the problem needs to be fixed on their end or not, or what the specific solution would be.

**Comment by kirkwaiblinger** [comment]
@ehoogeveen-medweb 

> If there's a particular change that plugin developers need to make to fix this problem for end users, I think adding that solution somewhere would be very helpful. I think some developers are taking a wait and see approach right now, not sure whether the problem needs to be fixed on their end or not, or what the specific solution would be.

I'd reiterate https://github.com/typescript-eslint/typescript-eslint/issues/11543#issuecomment-3240726967...

`TSESLint.FlatConfig.Parser`, `TSESLint.FlatConfig.Parser`, and `TSESLint.FlatConfig.Config` are not currently and are not foreseeably planned to be compatible with `defineConfig()` (and `RuleModule` will remain similarly incompatible). So, the short answer is: plugin authors can't use those types in their exports if they would like to be compatible with `defineConfig()`.

_We_ have solved this in typescript-eslint's own exports by giving our exports the following, rather absurd types (and not using the aforementioned `TSESLint.FlatConfig.*` types anymore):

https://github.com/typescript-eslint/typescript-eslint/blob/ef9173c56fa488737bff9ab463338ea65cb83373/packages/typescript-eslint/src/compatibility-types.ts#L1-L25

This solution is somewhat perverse in my opinion, and I don't really think it's worthy of recommendation, but it seemed to be the only way to satisfy `tseslint.config()` and `defineConfig()`, and for both users of eslint 8 or 9. If there is demand, perhaps we can entertain exporting these types for others to use, but... 😬 

In any case, other plugin authors may follow that pattern if they find it acceptable. But, beyond that, I'd say the best thing to do would be to seek advice from ESLint core on how they suggest authoring types for plugins.

---

The story of how we got in this crappy state is long and omitted for clarity, but suffice it to say that we don't find the current state of affairs satisfying either.

**Comment by reduckted** [comment]
> Is there anything that's actually actionable for us here? 

A function that takes in a strongly typed rule/plugin/parser/etc and returns a "weakly" typed equivalent might be useful. You can still use `typescript-eslint`'s types when creating those things, but then you end up with something that can be used with `defineConfig` (and hopefully in `taeslint.config` as well).

It would literally just return the given value and "cast" it, but having it as part of `typescript-eslint` would be convenient.

**Comment by bradzacher** [comment]
We should probably change our function to return the new types rather than the old types. That way plugin authors would get it "for free" just by upgrading?

**Comment by kirkwaiblinger** [comment]
> We should probably change our function...

Which function do you mean here?

**Comment by bradzacher** [comment]
The `RuleCreator` function -- https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/utils/src/eslint-utils/RuleCreator.ts

Currently it returns our types -- if we make it return the ESLint core types then a plugin created with those rules would be the same as if the author had retyped the plugin as an "ESLint core plugin", wouldn't it?

We can accept our types into the `RuleCreator` functions so that we can maintain control over the DevX -- but what it returns is irrelevant to the plugin authors -- so we can do the "transition" from our types to the core types there, I think.

**Comment by kirkwaiblinger** [comment]
Ah, I see what you're saying. It would require a bit of investigation. We couldn't do this for #10899 (for the plugin/parser types) because of eslint v8.57 compatibility requirements and problems with relevant eslint types not being exposed. Perhaps we'll get lucky with the rule module types, though 🤷.
 
The other point to keep in mind, though is that this doesn't help people who have written

```ts
const plugin: TSESLint.FlatConfig.Plugin = {
  rules: {
    // ...
  },
  // ...
}

export default plugin;
```

only people who have written 

```ts
export default {
  rules: {
    // ...
  },
  // ...
} satisfies TSESLint.FlatConfig.Plugin;
```

or similar.

**Comment by mikecentola** [comment]
Hello! I came across this searching because of some issues with the eslint-plugin-import-x compatibility with typescript-eslint, configuring it the default(ish) way. Instead of putting the additions at the top of the defineCconfig, I wanted them locked to the specific file type. 

```javascript
import tseslint from 'typescript-eslint';


export default defineConfig([

    globalIgnores([
        'out/*',
        'dist/*',
        'eslint.config.mjs',
    ]),

    {
        files: [
            'src/**/*.ts',
            'scripts/*.ts',
            'test/*.ts',
        ],
        ...js.configs.recommended,
        ...tseslint.configs.recommendedTypeChecked[0],

        languageOptions: {
            sourceType: 'module',
            parser: tseslint.parser,
            parserOptions: {
                ecmaVersion: 2024,
                ecmaFeatures: {
                    impliedStrict: true
                },
                projectService: true,
                tsconfigRootDir: import.meta.dirname,
            },
            globals: {
                ...globals.node,
                ...globals.mocha,
            }
        },

        plugins:{
            'import-x': importX, 
            '@typescript-eslint': tseslint.plugin
        },
        
        extends: [ 'import-x/flat/recommended'],
```

The things to note here is that I had to use ```...tseslint.configs.recommendedTypeChecked[0]``` to get it to resolve inside of the block with files.

In order to get the import-x working, I couldn't use their standard ```importX.flatConfigs.recommended``` but instead, I had to declare the plugin separately, and add the extends block. Additionally, that fired up some errors saying that the ```@typescript-eslint/.......``` was missing the plugin, so I redefined that plugin again with the import-x definition. I don't get any eslint errors now, but I still have to figure out how to get the resolving working.

Hope this helps a bit.

**-- Edit --**

Just needed to install ```eslint-import-resolver-typescript``` and add ```import-x/flat/typescript``` to the extends.  For anyone that comes across this...

**Comment by bradzacher** [comment]
@mikecentola your comment is unrelated to this thread.

**Comment by bradzacher** [comment]
> because of eslint v8.57 compatibility requirements

I think that we could probably just drop that requirement - without a breaking change - given that it was already a rare case.
Using eslint v8 plus flat configs was a rare thing - and if they were already using flat configs then the v9 migration would have been simple to action.

It might be worth prioritising a small major with the intent of solely addressing compatibility problems here. I.e. Release a new version which renames all of the types and swaps the types over so that we can "start afresh" where we need to.

**Comment by mikecentola** [comment]
> [@mikecentola](https://github.com/mikecentola) your comment is unrelated to this thread.

@bradzacher no worries. wasn't sure if my findings provided any insight to the issue at hand or at least a workaround. I was trying to dive deeper yesterday into it to see why there was that weird issue with import-x and using the import of typescript-eslint the way I did. Sorry about the confusion.

**Comment by kirkwaiblinger** [comment]
> > because of eslint v8.57 compatibility requirements
> 
> I think that we could probably just drop that requirement - without a breaking change - given that it was already a rare case. Using eslint v8 plus flat configs was a rare thing - and if they were already using flat configs then the v9 migration would have been simple to action.
>
> It might be worth prioritising a small major with the intent of solely addressing compatibility problems here. I.e. Release a new version which renames all of the types and swaps the types over so that we can "start afresh" where we need to.

Some thoughts:
- I think this would somewhat worsen the ESLint v8 to v9 migration path since it's useful to do the flat config migration first, and `tseslint.config()` comes in real handy for that, but the types would not necessarily work well for that anymore. On the other hand, most users will use the types for intellisense, not for type checking, so probably not a big problem there. And `tseslint.config()` is deprecated and mostly gone from the docs, anyway. 
- Would we compose on top of the ESLint v9 native types? I'm not sure that their types are neatly exported, especially in earlier versions of ESLint v9. So we might have to be intentional about what versions within ESLint v9 that we support types for. 
- I'm a bit worried that reworking the types will be a larger effort subject to scope creep, so it would be good to ensure this can be done with a limited set of changes... Otherwise I kind of prefer waiting till we can drop ESLint v8 🤷‍♂️

**Comment by bradzacher** [comment]
> since it's useful to do the flat config migration first

I'd disagree -- just cos you need to set an env var and also  you don't have any of the migration tooling that's builtin to ESLint.

> Otherwise I kind of prefer waiting till we can drop ESLint v8

Given ESLint v8 is still at ~50% usage -- that'll be a long time from now!
We shouldn't really just leave this open until then.

If we need to then we can push hard on ESLint core to do a better job of releasing their types so we (and other plugins) can depend on them without depending on the entire ESLint package.

**Comment by kirkwaiblinger** [comment]
no objection from me then 👍 Basically my understanding would be to 
1. Remove _type_ support guarantee for ESLint 8 flat configs
2. Keep _runtime_ support for ESLint ^8.57.0 (with eslintrc or flat config)
3. Ensure that 
   - `RuleCreator` returns rules that are compatible with ESLint 9
   - `FlatConfig.Parser` is assignable to ESLint parser type
   - `FlatConfig.Config` is assignable to ESLint config type
   - `FlatConfig.Plugin` is assignable to ESLint plugin type
4. We _might_ need to drop full type-level support for ESLint 8 plugins/rules authoring, too... For example, we might need to merge https://github.com/typescript-eslint/typescript-eslint/issues/11371 in order for the `RuleCreator` return type to work correctly.

**Comment by JoshuaKGoldberg** [comment]
Yeah this situation is a really irksome, painful point for the community. The plugins I (co-)maintain are stuck in the awkward limbo. We wouldn't want them to drop support for ESLint 8 so soon, so it would be painful for typescript-eslint to try to force that.

Kirk's plan sounds good to me. As long as ESLint 8 isn't _runtime_-broken, and there's a somewhat reasonable types story, I'm happy.

**Comment by Blackbaud-SteveBrush** [comment]
Have y'all thought about un-deprecating the `typescript-eslint` config utility? From the migration guide, it doesn't look like the `config` and `defineConfig` functions have feature parity (meaning, the `config` function offers different functionality from `defineConfig`), and appears to solve a valid use-case. We have a custom plugin and created a fairly complicated "recommended" config of our own, and getting 100+ teams to adopt the `defineConfig` function has a margin of error we're uncomfortable with.

**Comment by bradzacher** [comment]
@Blackbaud-SteveBrush that is unrelated to this issue. As per our contributing guide -- please file a new issue. https://typescript-eslint.io/contributing/issues#commenting

**Comment by 43081j** [comment]
Can we be perfectly clear what the recommended "hack" is for plugins meanwhile? 

I can see each plugin tackling this it's own way already. So can one of you possibly write an actual example diff or a before/after?

People finding this thread can then follow the advice at least

**Comment by Zamiell** [comment]
@43081j As a consumer of a plugin, you can do something like this:

```js
const myConfig = defineConfig({
  plugins: {
    // @ts-expect-error https://github.com/typescript-eslint/typescript-eslint/issues/11543
    foo: esLintPluginFoo,
  },
});
```

Then, when this issue is closed, you can simply remove the comment line.

Alternatively, you can refactor the plugin itself to not use `ESLintUtils.RuleCreator` at all. For an example of this, see [this custom `RuleCreator` function from the popular ESLint plugin `eslint-plugin-import-x`](https://github.com/un-ts/eslint-plugin-import-x/blob/master/src/utils/create-rule.ts). (Disclaimer: I don't know if this function actually passes `defineConfig`, so it might need some tweaking.)

**Comment by greedyivan** [comment]
We've been working on this issue and are about to open a PR. Since it is a breaking change that touches the types many of you annotate against, here is a short overview of the approach before the code lands.

**Why this matters.** Every plugin built with `ESLintUtils.RuleCreator` is runtime-compatible with `defineConfig()` and fails type-checking all the same — and this thread shows what that costs in practice: the canonical advice so far is a `@ts-expect-error` per plugin entry, plugins like import-x carry forked rule creators, angular-eslint ships its own boundary adapters. Meanwhile the typings still promise twelve rule-context members that the ESLint 10 runtime no longer delivers (#11371). The root is the split between our `TSESLint.FlatConfig.*` universe and the types ESLint ships and consumes today; everything above is a symptom of it.

**What we propose.** Anchor the public authoring types on `@eslint/core` — the runtime-agnostic package ESLint's own config entry points consume — added as a regular dependency of `@typescript-eslint/utils`. With that, `RuleCreator` outputs, `FlatConfig.Plugin/Parser/Config` and the configs `typescript-eslint` exports pass `defineConfig()` and `tseslint.config()` (including `extends` and spread composition) without casts. The twelve ghost members are removed, as a superset of #11386 — thanks @cseas, the PR carries your inventory forward. A migration guide covers the replacements for all twelve members, and a compatibility test matrix keeps both config entries green in CI.

**Why this shape.** A bridge type between the two universes was explored in #10899 and kept arriving at types too vague to be useful — the definitions diverged on shape, so the durable fix is one authority, not a smarter translator. `@eslint/core` is that authority: it is where the ecosystem is heading, and because it resolves from our own `node_modules`, consumers on any supported eslint major get one consistent answer. The change moves the type half of #11939 forward, keeps the precision this repo invests in — readonly `meta.defaultOptions` and the strict `RuleListener` are untouched, and the anchor needs exactly two narrow `satisfies`-guarded casts at the export boundary — and follows the plan agreed earlier in this thread: the type guarantee for ESLint 8 flat configs is lifted, runtime peer support is untouched.

The PR is a type-only change (one error-message string in `getParserServices` aside), with the full unit suites green. Details, the breaking-changes table and the file-by-file reading guide are in the PR description — feedback welcome, happy to adjust course on review.

**Comment by JoshuaKGoldberg** [comment]
Copying https://github.com/typescript-eslint/typescript-eslint/pull/12908#issuecomment-5750741779 here ~ there may be few needed breaking changes here. I don't have time to figure them out deeply. Some might be avoidable but I suspect at least a few won't be. This _might_ necessitate waiting for v9.

**`@typescript-eslint/utils`**

- `RuleListener`'s catch-all index signature went from `RuleFunction<never>` to `RuleFunction<any>`. Unannotated params on custom selectors (`'Foo[bar]'(node)`) are now `any` instead of `never`.
- `LooseRuleDefinition` is now an interface, and:
  - no longer accepts ESLint 8 function-style rules
  - `meta?: object | undefined` → `meta?: object` (only matters under `exactOptionalPropertyTypes`)
  - visitor values are `(...args: any[]) => void` instead of `Function`
- `FlatConfig.Plugin.processors`: `Partial<Record<…>>` → `Record<…> | undefined`. A `Partial`-typed value needs a cast.
- `SourceCode.VisitorKeys` values are `string[]` instead of `readonly string[]`. Anything passing readonly visitor keys where that type is expected errors — including our own rule-tester, which I had to touch.
- `SourceCode#applyInlineConfig` returns `{ configs, problems }` instead of `void`.

net effect: plugins/configs typed with `@types/eslint@8` no longer fit `FlatConfig.Plugin` / `tseslint.config()`. This is the "drop type support for ESLint 8 flat configs" thing we agreed on in the issue.

**`typescript-eslint`**

- `parser` is typed `FlatConfig.Parser` (a union) instead of `CompatibleParser`, so `tseslint.parser.parseForESLint` isn't directly accessible without narrowing anymore.
- `plugin` is typed `FlatConfig.Plugin` instead of `CompatiblePlugin`, so `tseslint.plugin.meta.name` goes from required to optional.
- `configs.*` are typed `FlatConfig.Config` / `ConfigArray` instead of `CompatibleConfig(Array)`. Mostly more precise, but it's a different type.
- All three of those are actually just going back to what they were before #11475 (8.40). `Compatible*` stay exported but `@deprecated`.

😬
